File: client/java/client/src/main/java/dev/angzarr/client/router/ProcessManagerResponse.java
package dev.angzarr.client.router;

import dev.angzarr.CommandBook;
import dev.angzarr.EventBook;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Response from process manager handlers.
 *
 * <p>Process managers produce:
 * <ul>
 *   <li>Commands (to send to other aggregates)</li>
 *   <li>Process events (to persist their own state)</li>
 *   <li>Facts (events to inject directly into other aggregates)</li>
 * </ul>
 */
public class ProcessManagerResponse {

    private final List<CommandBook> commands;
    private final EventBook processEvents;
    private final List<EventBook> facts;

    private ProcessManagerResponse(List<CommandBook> commands, EventBook processEvents, List<EventBook> facts) {
        this.commands = commands != null ? commands : Collections.emptyList();
        this.processEvents = processEvents;
        this.facts = facts != null ? facts : Collections.emptyList();
    }

    /**
     * Create an empty response (no commands, no process events, no facts).
     */
    public static ProcessManagerResponse empty() {
        return new ProcessManagerResponse(Collections.emptyList(), null, Collections.emptyList());
    }

    /**
     * Create a response with commands only.
     */
    public static ProcessManagerResponse withCommands(List<CommandBook> commands) {
        return new ProcessManagerResponse(new ArrayList<>(commands), null, Collections.emptyList());
    }

    /**
     * Create a response with process events only.
     */
    public static ProcessManagerResponse withProcessEvents(EventBook processEvents) {
        return new ProcessManagerResponse(Collections.emptyList(), processEvents, Collections.emptyList());
    }

    /**
     * Create a response with facts only.
     */
    public static ProcessManagerResponse withFacts(List<EventBook> facts) {
        return new ProcessManagerResponse(Collections.emptyList(), null, new ArrayList<>(facts));
    }

    /**
     * Create a response with commands and process events.
     */
    public static ProcessManagerResponse withBoth(List<CommandBook> commands, EventBook processEvents) {
        return new ProcessManagerResponse(new ArrayList<>(commands), processEvents, Collections.emptyList());
    }

    /**
     * Create a response with all fields.
     */
    public static ProcessManagerResponse withAll(List<CommandBook> commands, EventBook processEvents, List<EventBook> facts) {
        return new ProcessManagerResponse(
            new ArrayList<>(commands),
            processEvents,
            new ArrayList<>(facts)
        );
    }

    /**
     * Get the commands to send to other aggregates.
     */
    public List<CommandBook> getCommands() {
        return commands;
    }

    /**
     * Check if response has commands.
     */
    public boolean hasCommands() {
        return !commands.isEmpty();
    }

    /**
     * Get the process events to persist to PM's own domain.
     */
    public EventBook getProcessEvents() {
        return processEvents;
    }

    /**
     * Check if response has process events.
     */
    public boolean hasProcessEvents() {
        return processEvents != null;
    }

    /**
     * Get the facts to inject into other aggregates.
     */
    public List<EventBook> getFacts() {
        return facts;
    }

    /**
     * Check if response has facts.
     */
    public boolean hasFacts() {
        return !facts.isEmpty();
    }
}

File: client/csharp/Angzarr.Client/Router/ProcessManagerResponse.cs
namespace Angzarr.Client.Router;

/// <summary>
/// Response from process manager handlers.
/// Contains commands to send to other aggregates, events to persist
/// to the PM's own domain, and facts to inject into other aggregates.
/// </summary>
public class ProcessManagerResponse
{
    /// <summary>
    /// Commands to send to other aggregates.
    /// </summary>
    public List<Angzarr.CommandBook> Commands { get; set; } = new();

    /// <summary>
    /// Events to persist to the PM's own domain.
    /// </summary>
    public Angzarr.EventBook? ProcessEvents { get; set; }

    /// <summary>
    /// Facts to inject directly into other aggregates.
    /// </summary>
    public List<Angzarr.EventBook> Facts { get; set; } = new();

    /// <summary>
    /// Create an empty response.
    /// </summary>
    public static ProcessManagerResponse Empty() => new();

    /// <summary>
    /// Create a response with commands only.
    /// </summary>
    public static ProcessManagerResponse WithCommands(IEnumerable<Angzarr.CommandBook> commands) =>
        new() { Commands = commands.ToList() };

    /// <summary>
    /// Create a response with process events only.
    /// </summary>
    public static ProcessManagerResponse WithProcessEvents(Angzarr.EventBook processEvents) =>
        new() { ProcessEvents = processEvents };

    /// <summary>
    /// Create a response with facts only.
    /// </summary>
    public static ProcessManagerResponse WithFacts(IEnumerable<Angzarr.EventBook> facts) =>
        new() { Facts = facts.ToList() };

    /// <summary>
    /// Create a response with commands and process events.
    /// </summary>
    public static ProcessManagerResponse WithBoth(
        IEnumerable<Angzarr.CommandBook> commands,
        Angzarr.EventBook processEvents
    ) => new() { Commands = commands.ToList(), ProcessEvents = processEvents };

    /// <summary>
    /// Create a response with all fields.
    /// </summary>
    public static ProcessManagerResponse WithAll(
        IEnumerable<Angzarr.CommandBook> commands,
        Angzarr.EventBook processEvents,
        IEnumerable<Angzarr.EventBook> facts
    ) =>
        new()
        {
            Commands = commands.ToList(),
            ProcessEvents = processEvents,
            Facts = facts.ToList(),
        };
}
